File: fetch_videos.php
<?php

// The target URL for the login request
$loginUrl = 'https://gurumantrapsc.com/customer/login';
// The URL to fetch for a valid CSRF token
$homeUrl = 'https://gurumantrapsc.com/home';


// The phone number and password for the login attempt
$phoneNumber = '9809642422';
$password = 'rupesh123';

// Initialize a cURL session to get the CSRF token
$ch = curl_init($homeUrl);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_COOKIEJAR, 'cookie.txt');
curl_setopt($ch, CURLOPT_COOKIEFILE, 'cookie.txt');
curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
$html = curl_exec($ch);

// Now, get the login page to get a fresh CSRF token
curl_setopt($ch, CURLOPT_URL, $loginUrl);
$loginPageHtml = curl_exec($ch);

// Function to extract the CSRF token from the HTML content
function getCsrfToken($html) {
    if (preg_match('/<input type="hidden" name="_token" value="(.*?)?">/', $html, $matches)) {
        return $matches[1];
    }
    return false;
}

$csrfToken = getCsrfToken($loginPageHtml);

if ($csrfToken) {
    // Prepare the data to be sent in the POST request
    $data = [
        '_token' => $csrfToken,
        'phone_number' => $phoneNumber,
        'password' => $password,
    ];

    // Initialize a cURL session for the login request
    curl_setopt($ch, CURLOPT_URL, $loginUrl);
    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($data));
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);

    $response = curl_exec($ch);

    for ($i = 0; $i <= 10; $i++) {
        // The URL of the video to fetch
        $videoUrl = 'https://gurumantrapsc.com/purchase/video/' . $i;

        // Now, fetch the video page
        curl_setopt($ch, CURLOPT_URL, $videoUrl);
        curl_setopt($ch, CURLOPT_POST, false);
        $videoPage = curl_exec($ch);

        // Extract YouTube URL
        $youtubeUrl = '';
        if (preg_match('/<div class="plyr__video-embed" id="player">\s*<iframe.*?src="(.*?)".*?>/', $videoPage, $matches)) {
            $rawUrl = $matches[1];
            // Ensure we get a clean youtube.com/watch?v=... URL
            if(preg_match('/embed\/(.*?)(?:\?|$)/', $rawUrl, $id_match)){
                $youtubeUrl = 'https://www.youtube.com/watch?v=' . $id_match[1];
            }
        }

        // Extract Title
        $title = '';
        if (preg_match('/<h4 class="font-20 font-bold font-jakarta"> (.*?)<\/h4>/', $videoPage, $matches)) {
            $title = trim($matches[1]);
        }

        if ($youtubeUrl && $title) {
            echo $i . "=" . $youtubeUrl . ":" . $title . "\n";
        }
    }

} else {
    echo "Could not find CSRF token on the live page.";
}

curl_close($ch);

?>

File: fetch_videos.py
import requests
from bs4 import BeautifulSoup
import time
import os
import threading
from flask import Flask, send_from_directory

# Configuration
LOGIN_URL = 'https://gurumantrapsc.com/customer/login'
HOME_URL = 'https://gurumantrapsc.com/home'
PHONE_NUMBER = os.environ.get('GURU_PHONE')     # set in Render dashboard
PASSWORD = os.environ.get('GURU_PASS')          # set in Render dashboard
RESULTS_FILE = 'education_latest.txt'

# Flask App setup
app = Flask(__name__)

@app.route('/')
@app.route('/ping')
def ping():
    return 'PONG', 200

@app.route('/education_latest.txt')
def serve_education_file():
    return send_from_directory(os.getcwd(), RESULTS_FILE, as_attachment=False)

def get_csrf_token(session, url):
    try:
        response = session.get(url)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, 'html.parser')
        token_tag = soup.find('input', {'name': '_token'})
        if token_tag:
            return token_tag['value']
    except requests.exceptions.RequestException as e:
        print(f"Error fetching CSRF token: {e}")
    return None

def login(session, csrf_token):
    payload = {
        '_token': csrf_token,
        'phone_number': PHONE_NUMBER,
        'password': PASSWORD,
    }
    try:
        response = session.post(LOGIN_URL, data=payload)
        response.raise_for_status()
        if "customer/dashboard" in response.url:
            print("Login successful.")
            return True
        else:
            print("Login failed.")
            return False
    except requests.exceptions.RequestException as e:
        print(f"Error during login: {e}")
        return False

def fetch_video_info(session, video_id):
    video_url = f'https://gurumantrapsc.com/purchase/video/{video_id}'
    try:
        headers = {
            'Referer': 'https://gurumantrapsc.com/customer/dashboard'
        }
        response = session.get(video_url, headers=headers)
        response.raise_for_status()

        soup = BeautifulSoup(response.text, 'html.parser')

        youtube_url = None
        iframe = soup.find('iframe')
        if iframe and iframe.has_attr('src') and 'youtube.com/embed/' in iframe['src']:
            video_id_youtube = iframe['src'].split('/embed/')[1].split('?')[0]
            youtube_url = f"https://www.youtube.com/watch?v={video_id_youtube}"

        title = None
        header = soup.find('div', class_='dashboard-header')
        if header and header.find('p'):
            title = header.find('p').text.strip()
        elif soup.title:
            title = soup.title.string.replace("| Gurumantra", "").strip()

        return youtube_url, title

    except requests.exceptions.RequestException as e:
        print(f"Error fetching video {video_id}: {e}")
        return None, None

def scrape_videos():
    existing_results = set()

    if not os.path.exists(RESULTS_FILE):
        with open(RESULTS_FILE, 'w') as f:
            pass

    with requests.Session() as session:
        session.headers.update({
            'User-Agent': 'Mozilla/5.0',
            'Accept': '*/*',
        })

        csrf_token = get_csrf_token(session, HOME_URL)
        if not csrf_token:
            print("Could not get CSRF token.")
            return

        if not login(session, csrf_token):
            print("Login failed.")
            return

        current_video_id = 6344
        last_found_video_id = 0
        consecutive_not_found = 0

        while True:
            try:
                with open(RESULTS_FILE, 'r') as f:
                    for line in f:
                        if '=' in line:
                            existing_results.add(line.split('=')[0])
            except FileNotFoundError:
                pass

            if str(current_video_id) in existing_results:
                print(f"SKIPPED: {current_video_id}")
                current_video_id += 1
                continue

            print(f"Fetching video ID: {current_video_id}")
            youtube_url, title = fetch_video_info(session, current_video_id)

            if youtube_url and title:
                result_line = f"{current_video_id}={youtube_url}:{title}\n"
                print(f"FOUND: {result_line}", end='')
                with open(RESULTS_FILE, 'a') as f:
                    f.write(result_line)
                last_found_video_id = current_video_id
                consecutive_not_found = 0
            else:
                consecutive_not_found += 1
                print(f"Not found: {current_video_id}")
                if consecutive_not_found >= 100:
                    print(f"100 not found. Restarting from {last_found_video_id}")
                    current_video_id = last_found_video_id
                    consecutive_not_found = 0

            print("------------------------------------")
            time.sleep(1)
            current_video_id += 1

def auto_restart_scraper():
    while True:
        try:
            scrape_videos()
        except Exception as e:
            print(f"[AutoRestart] Scraper crashed: {e}")
        print("Restarting in 5 seconds...")
        time.sleep(5)

if __name__ == "__main__":
    scraper_thread = threading.Thread(target=auto_restart_scraper)
    scraper_thread.daemon = True
    scraper_thread.start()

    port = int(os.environ.get('PORT', 5000))
    app.run(host='0.0.0.0', port=port)
